docs/SystemConfig.md
C-SkillSys offers several patches to allow game-designers to configure some of the underlying logic of the game without touching the code.

## Designer config

![image](./gfx/SystemConfig1.png)

- `equip_skill_en`

	Three-Houses style skill equipment in prepscreen is enabled to Allow units equip skills in prepscreen. Otherwise the entrance of equip option in prepscreen will be removed.

- `gen_new_scroll`

	When unit uses skill scroll who's skill equipable slot has been fullfilled, c-skillsys can choose one of the following options:

	- `0`: Three houses style: just register the new skill into the character's list of learned skills. And then he may equip that skill in prepscreen.

	- `1`: Old style: allow players remove an existing skill on unit and generated a new skill scroll.

- `max_equipable_skill`

	The total number of skills that can be dynamically disassembled for each character (0-7).

- `remove_move_path`

	Disable move-path calculation and display.

	Since move-path can display no more than 20 steps in vanilla, with the increase of unit's mobility, this limitation may cause risk of overflow error. Thus we can directly remove the move-path to avoid overflow.

- `use_chinese_character`

	![image](./gfx/SystemConfig1_ChChar.png)

	Display Chinese affin character in statscreen unit page.

- `kernel_tutorial_level`

	Optional display kernel system tutorial box.

	- `0`: Not show any kernel tutorial
	- `1`: Only show kernel tutorial in tutorial mode
	- `2`: show kernel tutorial at anytime

- `combo_attack_en`

	Enable FE-Engage style combo-attack.

- `battle_surrend_en`

	Enable battle status loss when unit is surrendered.

- `hit_decrease_on_range`

	Enable FE-ThreeHouses style hit rate loss for far-ranged attack.

- `debug_autoload_skills`

	**Debug option**, fullfill the unit learned-skills list so that he can freely select skills to equip in prepscreen.

- `guaranteed_lvup`

	If no status increases during level-up, kernel may retry 10 times with a 10% growth bonus.

- `lvup_mode_tutorial`

	Level up mode in tutorial mode,

	- `0`: GBA style lvup, same as vanilla
	- `1`: 3DS style lvup, uncontrollable random lvup
	- `2`: fixed lvup
	- `3`: 100% growth
	- `4`: 0% growth

- `lvup_mode_normal`

	Level up mode in normal mode, same as `lvup_mode_tutorial`.

- `lvup_mode_hard`

	Level up mode in hard mode, same as `lvup_mode_tutorial`.

- `unit_page_style`

	Display style of statscreen page1.

	- `0`: ![image](./gfx/SystemConfig1_UnitPage1.png)
	- `1`: ![image](./gfx/SystemConfig1_UnitPage2.png)

- `skill_page_style`

	Display style of statscreen page3.

	- `0`: ![image](./gfx/SystemConfig1_SkillPage1.png)
	- `1`: ![image](./gfx/SystemConfig1_SkillPage2.png)

- `gaiden_magic_en`

	Enable gaiden style black/white magic system.

- `gaiden_magic_must_be_magic`

	Weapons in gaiden B/W.mag selection can also be non-magic items.

- `gaiden_magic_requires_wrank`

	If unit can use one magic, he may use that weapon ragardless the weapon-rank level requirement.

- `gaiden_magic_ai_en`

	Enemies can also use gaiden magic on AI phase.

- `gaiden_magic_ext_conf_en`

	**Wizardry option**, enable `gGaidenChaxConfigs` judgement.

- `auto_narrow_font`

	Try convert ascii to **narrow-fonts** on displaying skill descriptions, skill names and menu items.

- `skill_sub_menu_width`

	Width of submenu after typing "Skills" option in action menu.

- `wrank_bonux_rtext_auto_gen`

	Display weapon rank battle bonus info in statscreen if it holds.

- `enemy_can_combo_attack`

	Enemy can also participant in combo-attack.

- `menu_skill_disp_msg_en_n`

	Menu skill display option, see [SkillSys](./SkillSys.md) section 7.1.

- `banim_switcher_en`

	Enable patch "Custom banim switcher", see [BanimFeatures](./BanimFeatures.md) section 1.

- `max_level`

	Max level for units (0-25).

- `max_level_record`

	Max total level for units (current level + hidden level, 0-80). See SkillSys.md section 3.2.

## Designer config battle

![image](./gfx/SystemConfig2.png)

- `ATK percentage in AS calc`

	Attack speed decay: (weight - con) --> (weight - (con + atk * `[20%]`))

- `HP drain percentgae for nosferatu`

	Default hp-drain percentage: 100% --> `[50%]`.(Lower priority than [kernel item config table](../Data/BattleSys/WeaponHpDrain.c))

- `AVO deacy for riders indorr`

	Riders may suffer (avo -`[20%]`) indoor.

- `CRIT damage correction rate`

	Critical attack may cause `[200%]` damage.

- `Critical rate bonus for cavalry`

	Cavalry class suffer crit `[-0%]`

- `Critical rate bonus on attributes bit`

	If bit "Myrmidon/SwordMaster" in Ability3 is set, get crit `[+15%]`

- `combo_base_damage`
- `combo_additional_damage_en`
- `combo_additional_damage_perc`

	For combo-attack, battle damage calc:

	`combo_base_damage` + [atk - def] * `combo_additional_damage_perc`% * !!`combo_additional_damage_en`

- `battle_followup_speed_threshold`

	Configurable value for vanilla `BATTLE_FOLLOWUP_SPEED_THRESHOLD`.

## Skill misc infos

![image](./gfx/SystemConfig3.png)

Allow designers to customize some numerical attributes of skills. The meaning of each value is related to the skill. Please note that you need to be careful when modifying each of these values, as this may cause bugs.


Patches/PATCH_DeignerConfig.txt
NAME=[CSKILL_K dev] designer config
AUTHOR=Mokha(悠木晓缺)

IF:0xB2A604=0x43 0x53 0x4B 0x49 0x4C 0x4C 0x53 0x59 0x53 0x5F 0x4B 0x2D 0x64 0x65 0x76 0x00

TYPE=STRUCT
TAG=#ENGINE

DATASIZE=32
DATACOUNT=1

POINTER=0xB2A680
B0:$COMBO combo.bool.txt=equip_skill_en
B1:DECIMAL=max_equipable_skill (0-7)
B2:$COMBO combo.bool.txt=remove_move_path
B3:$COMBO combo.bool.txt=use_chinese_character
B4:$COMBO combo.ktut_level.txt=kernel_tutorial_level
B5:$COMBO combo.bool.txt=combo_attack_en
B6:$COMBO combo.bool.txt=battle_surrend_en
B7:$COMBO combo.bool.txt=hit_decrease_on_range
B8:$COMBO combo.bool.txt=debug_autoload_skills
B9:$COMBO combo.bool.txt=guaranteed_lvup
B10:$COMBO combo.lvupmode.txt=lvup_mode_tutorial
B11:$COMBO combo.lvupmode.txt=lvup_mode_normal
B12:$COMBO combo.lvupmode.txt=lvup_mode_hard
B13:$COMBO combo.PageStyle_Unit.txt=unit_page_style
B14:$COMBO combo.PageStyle_Skill.txt=skil_page_style

INFO=Editor: gKernelDesigerConfig


Tools/EventAssembler/Advance/FE8/fe8u-cskillsys-kernel/Tools/EventAssembler/Nintendo/Game Boy Advance/-Lex Talinus-/FE8RBeta/lt_engine/lt_engine/app/engine/settings.py
from app.constants import WINWIDTH, WINHEIGHT

from app.engine import config as cf
from app.engine.sprites import SPRITES
from app.engine.fonts import FONT
from app.engine.sound import get_sound_thread
from app.engine.input_manager import get_input_manager
from app.engine.state import State
from app.engine import engine, background, banner, menus, settings_menu, base_surf, text_funcs
from app.engine.game_state import game
from app.engine.fluid_scroll import FluidScroll

controls = {'key_SELECT': engine.subsurface(SPRITES.get('buttons'), (0, 66, 14, 13)),
            'key_BACK': engine.subsurface(SPRITES.get('buttons'), (0, 82, 14, 13)),
            'key_INFO': engine.subsurface(SPRITES.get('buttons'), (1, 149, 16, 9)),
            'key_AUX': engine.subsurface(SPRITES.get('buttons'), (1, 133, 16, 9)),
            'key_START': engine.subsurface(SPRITES.get('buttons'), (0, 165, 33, 9)),
            'key_LEFT': engine.subsurface(SPRITES.get('buttons'), (1, 4, 13, 12)),
            'key_RIGHT': engine.subsurface(SPRITES.get('buttons'), (1, 19, 13, 12)),
            'key_DOWN': engine.subsurface(SPRITES.get('buttons'), (1, 34, 12, 13)),
            'key_UP': engine.subsurface(SPRITES.get('buttons'), (1, 50, 12, 13))}
control_order = ('key_SELECT', 'key_BACK', 'key_INFO', 'key_AUX', 'key_LEFT', 'key_RIGHT', 'key_UP', 'key_DOWN', 'key_START')

config = [('animation', ['Always', 'Your Turn', 'Combat Only', 'Never'], 0),
          ('screen_size', [1, 2, 3, 4, 5, 6], 18),
          ('display_fps', bool, 2),
          ('battle_bg', bool, 17),
          ('unit_speed', list(reversed(range(15, 180, 15))), 1),
          ('text_speed', cf.text_speed_options, 2),
          # ('cursor_speed', list(reversed(range(32, 160, 16))), 8),
          ('mouse', bool, 19),
          ('show_terrain', bool, 7),
          ('show_objective', bool, 6),
          ('autocursor', bool, 13),
          ('hp_map_team', ['All', 'Ally', 'Enemy'], 10),
          ('hp_map_cull', ['None', 'Wounded', 'All'], 10),
          ('music_volume', [x/10.0 for x in range(0, 11, 1)], 15),
          ('sound_volume', [x/10.0 for x in range(0, 11, 1)], 16),
          ('talk_boop', bool, 16),
          ('show_bounds', bool, 7),
          ('grid_opacity', [int(255 * x / 10.0) for x in range(11)], 7),
          ('autoend_turn', bool, 14),
          ('confirm_end', bool, 14),
          ('display_hints', bool, 3)]

config_icons = [engine.subsurface(SPRITES.get('settings_icons'), (0, c[2] * 16, 16, 16)) for c in config]

class SettingsMenuState(State):
    name = 'settings_menu'
    in_level = False
    header_width = 112

    def start(self):
        self.fluid = FluidScroll(128)
        self.bg = background.create_background('settings_background')
        # top_menu_left, top_menu_right, config, controls, get_input
        self.state = 'top_menu_left'

        control_options = control_order
        control_icons = [controls[c] for c in control_options]
        self.controls_menu = settings_menu.Controls(None, control_options, 'menu_bg_base', control_icons)
        self.controls_menu.takes_input = False

        config_options = [(c[0], c[1]) for c in config]
        self.config_menu = settings_menu.Config(None, config_options, 'menu_bg_base', config_icons)
        self.config_menu.takes_input = False

        self.top_cursor = menus.Cursor()

        game.state.change('transition_in')
        return 'repeat'

    @property
    def current_menu(self):
        if self.state in ('top_menu_left', 'config'):
            return self.config_menu
        else:
            return self.controls_menu

    def handle_mouse(self):
        mouse_position = get_input_manager().get_mouse_position()
        if mouse_position:
            mouse_x, mouse_y = mouse_position
            top_left_rect = (4, 4, self.header_width, 24)
            top_right_rect = (WINWIDTH//2 + 4, 4, self.header_width, 24)
            # Test left rect
            x, y, width, height = top_left_rect
            if x <= mouse_x <= x + width and y <= mouse_y <= y + height:
                self.current_menu.takes_input = False
                self.state = 'top_menu_left'
                return
            # Test right rect
            x, y, width, height = top_right_rect
            if x <= mouse_x <= x + width and y <= mouse_y <= y + height:
                self.current_menu.takes_input = False
                self.state = 'top_menu_right'
                return
            current_idxs, current_option_rects = self.current_menu.get_rects()
            for idx, option_rect in zip(current_idxs, current_option_rects):
                x, y, width, height = option_rect
                if x <= mouse_x <= x + width and y <= mouse_y <= y + height:
                    if self.state in ('top_menu_left', 'config'):
                        self.state = 'config'
                    else:
                        self.state = 'controls'
                    self.current_menu.takes_input = True
                    self.current_menu.move_to(idx)
                    return

    def take_input(self, event):
        first_push = self.fluid.update()
        directions = self.fluid.get_directions()

        if self.state == 'get_input':
            if event == 'BACK':
                get_sound_thread().play_sfx('Select 4')
                self.state = 'controls'
                get_input_manager().set_change_keymap(False)
            elif event == 'NEW':
                get_sound_thread().play_sfx('Select 1')
                self.state = 'controls'
                selection = self.current_menu.get_current()
                cf.SETTINGS[selection] = get_input_manager().unavailable_button
                get_input_manager().set_change_keymap(False)
                get_input_manager().update_key_map()
            elif event:
                get_sound_thread().play_sfx('Select 4')
                self.state = 'controls'
                get_input_manager().set_change_keymap(False)
                text = 'Invalid Choice!'
                game.alerts.append(banner.Custom(text))
                game.state.change('alert')

        elif self.state in ('top_menu_left', 'top_menu_right'):
            self.handle_mouse()
            if event == 'DOWN' or event == 'SELECT':
                get_sound_thread().play_sfx('Select 6')
                if self.state == 'top_menu_left':
                    self.state = 'config'
                else:
                    self.state = 'controls'
                self.current_menu.takes_input = True
            elif event == 'LEFT':
                if self.state == 'top_menu_right':
                    get_sound_thread().play_sfx('Select 6')
                    self.state = 'top_menu_left'
            elif event == 'RIGHT':
                if self.state == 'top_menu_left':
                    get_sound_thread().play_sfx('Select 6')
                    self.state = 'top_menu_right'
            elif event == 'BACK':
                self.back()

        else:
            self.handle_mouse()
            if 'DOWN' in directions:
                get_sound_thread().play_sfx('Select 6')
                self.current_menu.move_down(first_push)
            elif 'UP' in directions:
                get_sound_thread().play_sfx('Select 6')
                if self.current_menu.get_current_index() <= 0:
                    self.current_menu.takes_input = False
                    if self.state == 'config':
                        self.state = 'top_menu_left'
                    else:
                        self.state = 'top_menu_right'
                else:
                    self.current_menu.move_up(first_push)
            elif 'LEFT' in directions:
                get_sound_thread().play_sfx('Select 6')
                self.current_menu.move_left()
                if self.current_menu.get_current_option().name in ('music_volume', 'sound_volume'):
                    self.update_sound()
            elif 'RIGHT' in directions:
                get_sound_thread().play_sfx('Select 6')
                self.current_menu.move_right()
                if self.current_menu.get_current_option().name in ('music_volume', 'sound_volume'):
                    self.update_sound()

            elif event == 'BACK':
                self.back()

            elif event == 'SELECT':
                if self.state == 'controls':
                    get_sound_thread().play_sfx('Select 1')
                    self.state = 'get_input'
                    get_input_manager().set_change_keymap(True)
                elif self.state == 'config':
                    get_sound_thread().play_sfx('Select 6')
                    self.current_menu.move_next()
                    if self.current_menu.get_current_option().name in ('music_volume', 'sound_volume'):
                        self.update_sound()

    def back(self):
        get_sound_thread().play_sfx('Select 4')
        cf.save_settings()
        self.update_sound()
        # if game.cursor is not None:
        #     game.cursor.fluid.update_speed(cf.SETTINGS['cursor_speed'])
        game.state.change('transition_pop')

    def update_sound(self):
        get_sound_thread().set_music_volume(cf.SETTINGS['music_volume'])
        get_sound_thread().set_sfx_volume(cf.SETTINGS['sound_volume'])

    def update(self):
        self.current_menu.update()
        self.top_cursor.update()

    def draw_top_menu(self, surf):
        bg = base_surf.create_base_surf(self.header_width, 24, 'menu_bg_clear')
        offset = (WINWIDTH // 2 - self.header_width) // 2
        surf.blit(bg, (offset, 4))
        surf.blit(bg, (WINWIDTH//2 + offset, 4))
        if self.current_menu is self.config_menu:
            FONT['text-yellow'].blit_center('Config', surf, (offset + self.header_width//2, 8))
            FONT['text-grey'].blit_center('Controls', surf, (WINWIDTH//2 + offset + self.header_width//2, 8))
            if self.state in ('top_menu_left', 'top_menu_right'):
                self.top_cursor.draw(surf, self.header_width//2 - 16, 8)
        else:
            FONT['text-grey'].blit_center('Config', surf, (offset + self.header_width/2, 8))
            FONT['text-yellow'].blit_center('Controls', surf, (WINWIDTH//2 + offset + self.header_width//2, 8))
            if self.state in ('top_menu_left', 'top_menu_right'):
                self.top_cursor.draw(surf, WINWIDTH//2 + 2 + self.header_width//2 - 16, 8)

    def draw_info_banner(self, surf):
        height = 16
        bg = base_surf.create_base_surf(WINWIDTH + 16, height, 'menu_bg_clear')
        surf.blit(bg, (-8, WINHEIGHT - height))
        if self.state == 'top_menu_left':
            text = 'config_desc'
        elif self.state == 'top_menu_left':
            text = 'controls_desc'
        elif self.state == 'config':
            idx = self.config_menu.get_current_index()
            text = config[idx][0] + '_desc'
        elif self.state == 'get_input':
            text = 'get_input_desc'
        else:
            text = 'keymap_desc'
        text = text_funcs.translate(text)
        FONT['text'].blit_center(text, surf, (WINWIDTH//2, WINHEIGHT - height))

    def draw(self, surf):
        if self.bg:
            self.bg.draw(surf)
        else:
            # settings menu shouldn't be transparent
            surf.blit(SPRITES.get('bg_black'), (0, 0))

        self.draw_top_menu(surf)
        if self.state == 'get_input':
            self.current_menu.draw(surf, True)
        else:
            self.current_menu.draw(surf)
        self.draw_info_banner(surf)

        return surf

    def finish(self):
        # Just to make sure!
        get_input_manager().set_change_keymap(False)
